tests/testsuite/publish.rs

```diff
@@ -1078,8 +1086,8 @@ Check for a source-replacement in .cargo/config.
         .with_status(101)
         .with_stderr(
             "\
-[ERROR] dir [..]/foo/vendor does not support API commands.
-Check for a source-replacement in .cargo/config.
+[ERROR] crates-io is replaced with non-remote-registry source dir [..]/foo/vendor;
+include `--registry crates-io` to publish to crates.io
```

Comment:
I don't know how to test that with the existing framework. Passing `--registry crates-io` will ensure that it talks directly to real crates.io

Reply:
@ehuss do you have any insight on how we should test this? `--registry crates-io` is supposed use real crates.io, so there's not a good way to validate in the existing test framework of integration-style tests.

We could:
* Have an environment variable to override the URL for crates.io that would be set for these tests
* Intercept the HTTP requests via a proxy (this will be difficult given that it's https)

Reply:
Hm, that's really a sticky issue. I have a vague memory that in the past we have avoided a test-only env var to change the crates.io URL, as it is likely someone will discover it and use it. It's also a bit uncomfortable to add separate code paths for testing since it isn't testing the actual code paths.

Unfortunately I can't think of any other good solutions. I don't think a proxy is feasible, as I believe curl will use CONNECT to tunnel through, which I don't think can be intercepted. It would need to be a mitm proxy, and I don't think that's possible here.

I suppose if the env var is similar to what you have used here it might be ok.

Reply:
I used an environment variable to allow tests to override the crates.io URL.

I added a new `source_replacement.rs` test file that tests these cases (including --registry crates-io).